### PR TITLE
Introduce storybook to represent react components.

### PR DESCRIPTION
**Pull request by ohadlevy**
to use:

npm install storybook

npm run storybook

### CODE REVIEW

File: config/webpack.common.config.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
+'use strict';
+
+require('es6-promise').polyfill(); // needed for compatibility with older node versions
+
+var path = require('path');
```
Comment (gailsteiger) [inline config/webpack.common.config.js]: The webpack config files are not processed by babel. Add 

``` js
/* eslint-disable no-var */
```

at the top of the file to disable the no-var rule. 'var' declarations should still be at the top of the function scope.

File: webpack/assets/javascripts/react_app/components/notifications/Toast.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import {Button} from 'react-bootstrap';
+
+const Toast = (props) => {
+  let type = (props.type || 'success');
```
Comment (gailsteiger) [inline webpack/assets/javascripts/react_app/components/notifications/Toast.js]: use const here instead of let

**Comment by ohadlevy** [comment]
@gailsteiger updated all of hound concerns and your comments.

File: config/webpack.common.config.js (hunk 1 above)
Comment (ohadlevy) [reply config/webpack.common.config.js]: added the whole file to eslint ignore.

Comment (gailsteiger) [reply config/webpack.common.config.js]: I think that's extreme. I think that it would be better to lint the file and disable specific rules that aren't relevant.

**Comment by ohadlevy** [comment]
I was just following the same as its done to webpack configuration it self
currently. /cc @tbrisker as in that case I would assume that we should lint
the current webpack configuration?

**Comment by gailsteiger** [comment]
@ohadlevy I was surprised to find webpack config  in the ignore list. I think we should lint them both.

**Comment by ohadlevy** [comment]
rebased, @tbrisker or @dLobatog will you have a minute to review? thanks!

**Comment by ohadlevy** [comment]
rebased.

File: .storybook/head.html [hunk 3]
```diff
@@ -0,0 +1,14 @@
+<link href='https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css' rel='stylesheet' media="screen, print">
+<link href='https://cdnjs.cloudflare.com/ajax/libs/patternfly/3.10.0/css/patternfly.min.css' rel='stylesheet' media="screen, print">
+<link href='https://cdnjs.cloudflare.com/ajax/libs/patternfly/3.10.0/css/patternfly-additions.min.css' rel='stylesheet' media="screen, print">
+<script src='https://cdnjs.cloudflare.com/ajax/libs/jquery/1.11.3/jquery.min.js'></script>
+<script src='https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js'></script>
+<script src='https://cdnjs.cloudflare.com/ajax/libs/d3/4.3.0/d3.min.js'></script>
+<script src='https://cdnjs.cloudflare.com/ajax/libs/c3/0.4.11/c3.min.js'></script>
+<!-- <script src='https://cdnjs.cloudflare.com/ajax/libs/jed/1.1.1/jed.min.js'></script> -->
+<!-- <script src='https://cdnjs.cloudflare.com/ajax/libs/patternfly/3.10.0/js/patternfly.min.js'></script> -->
```
Comment (tbrisker) [inline .storybook/head.html]: any reason we are using CDN here rather then the versions already available in our assets? might lead to unexpected differences in layout if we have different versions.

File: .eslintignore [hunk 4]
```diff
@@ -2,3 +2,4 @@
 app/**
 vendor/**
 config/webpack.config.js
+config/webpack.common.config.js
```
Comment (tbrisker) [inline .eslintignore]: I think we should just ignore `config/**`, no point linting any of it (since we lint with es6 rules etc and config has to be legacy)

File: config/webpack.common.config.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
+/* eslint-disable no-var */
```
Comment (tbrisker) [inline config/webpack.common.config.js]: we don't lint this, no point in adding comments for the linter

File: config/webpack.config.js [hunk 6]
```diff
@@ -1,17 +1,15 @@
+/* eslint-disable no-var */
```
Comment (tbrisker) [inline config/webpack.config.js]: ditto

File: config/webpack.config.js [hunk 7]
```diff
@@ -1,17 +1,15 @@
...
-var webpack = require('webpack');
-var StatsPlugin = require('stats-webpack-plugin');
-var ExtractTextPlugin = require("extract-text-webpack-plugin");
+var GlobalConfig = require('../config/webpack.common.config');
 
 // must match config.webpack.dev_server.port
 var devServerPort = 3808;
 
 // set TARGETNODE_ENV=production on the environment to add asset fingerprints
 var production = process.env.RAILS_ENV === 'production' || process.env.NODE_ENV === 'production';
```
Comment (tbrisker) [inline config/webpack.config.js:14]: this is duplicated in webpack.common.config.js. Should we also split develop and production into different config files?

File: webpack/assets/javascripts/react_app/components/charts/Chart.js [hunk 8]
```diff
@@ -30,7 +30,9 @@ class Chart extends React.Component {
   }
 
   componentWillUnmount() {
-    this.chart = this.chart.destroy();
+    if (this.hasData()) {
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/charts/Chart.js]: is this change related?

File: webpack/assets/javascripts/react_app/components/common/Search.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
+import React, { Component } from 'react';
+import helpers from '../../common/helpers';
+import {Form, FormGroup, ControlLabel, FormControl, Button} from 'react-bootstrap';
+
+class Search extends Component {
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/common/Search.js]: I would prefer not introducing new component in this PR as the subject is introducing storybook.

File: webpack/assets/javascripts/react_app/components/notifications/Toast.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import {Button} from 'react-bootstrap';
+
+const ToastIcons = {
+  success: 'ok',
+  danger: 'error-circle-o',
+  warning: 'warning-triangle-o',
+  'default': 'info'
+};
+
+const Toast = (props) => {
```
Comment (tbrisker) [inline webpack/assets/javascripts/react_app/components/notifications/Toast.js]: ditto

File: config/webpack.config.js (hunk 7 above)
Comment (tbrisker) [reply config/webpack.config.js:14]: also, if we have multiple config files, maybe place them all under `config/webpack/`?

**Review by tbrisker — CHANGES_REQUESTED** [review]

File: .storybook/head.html (hunk 3 above)
Comment (ohadlevy) [reply .storybook/head.html]: once they are moved into npm packages, we can use it, this is kind of quick and dirty approach to get it going

File: .eslintignore (hunk 4 above)
Comment (ohadlevy) [reply .eslintignore]: I'm in favor of linting it, but I don't want to do it in this pr, regardless, i can change it to ```*``` but I think we should not ignore by default.

File: config/webpack.common.config.js (hunk 5 above)
Comment (ohadlevy) [reply config/webpack.common.config.js]: its my editor which complains :)

File: config/webpack.config.js (hunk 7 above)
Comment (ohadlevy) [reply config/webpack.config.js:14]: or maybe webpack/config

File: webpack/assets/javascripts/react_app/components/charts/Chart.js (hunk 8 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/charts/Chart.js]: semi related, most likely should be done in refactoring to charts components to be able to be used in storybook... /cc @gailsteiger

File: webpack/assets/javascripts/react_app/components/common/Search.js (hunk 9 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/common/Search.js]: agree, I'll remove it.

File: config/webpack.config.js (hunk 7 above)
Comment (tbrisker) [reply config/webpack.config.js:14]: whichever, as long as we paint the bikeshed green :smile:

File: webpack/stories/index.js [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+      modalConfig={{}}
+      noDataMsg={'No data here'}
+      id="target"
+      status="RESOLVED"
+    />
+  )
+);
+
+storiesOf('Power Status', module)
+  .add('Loading', () => (
```
Comment (ohadlevy) [inline webpack/stories/index.js:56]: @tbrisker I removed the other components and replace them with power status instead.

@gailsteiger from some reason, the imported css of the PowerStatus.js is not visible in storybook, i suspect something in webpack, but I was not able to identify it, could you please have a look?

Comment (gailsteiger) [reply webpack/stories/index.js:56]: @ohadlevy Seems to work if you remove the '.storybook/webpack.config.js' file.

**Comment by ohadlevy** [comment]
@tbrisker removed all webpack changes (seems like storybook defaults were working fine), added power status stories.

File: .storybook/head.html (hunk 3 above)
Comment (ohadlevy) [reply .storybook/head.html]: do you have an alternative idea? (I'm already importing the asset pipieline application.js - however webpack ignores the //= require statements)

File: .storybook/head.html [hunk 12]
```diff
@@ -0,0 +1,15 @@
+<link href='https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css' rel='stylesheet' media="screen, print">
+<link href='https://cdnjs.cloudflare.com/ajax/libs/patternfly/3.10.0/css/patternfly.min.css' rel='stylesheet' media="screen, print">
+<link href='https://cdnjs.cloudflare.com/ajax/libs/patternfly/3.10.0/css/patternfly-additions.min.css' rel='stylesheet' media="screen, print">
+<script src='https://cdnjs.cloudflare.com/ajax/libs/jquery/1.12.4/jquery.min.js'></script>
```
Comment (ohadlevy) [inline .storybook/head.html:4]: updated to a current jquery version.

I did try to use npm jquery package, however, since bootstrap is not in npm i was unable, as jquery needs to load prior to bootstrap.
the way that bootstrap works is by adding the bundle inside the body vs the head section.

**Comment by ohadlevy** [comment]
@tbrisker @dLobatog anything else needed to get this moving? thanks!

**Review by tbrisker — APPROVED** [review]

**Comment by tbrisker** [comment]
Ready to merge on my side, going to give this a few hours before merging in case @dLobatog or anyone else have further comments.

**Comment by ohadlevy** [comment]
@tbrisker :+1:

**Comment by tbrisker** [comment]
Merged, Thanks @ohadlevy !
